### ISSUE TITLE
Bag already exists

### ISSUE THREAD

**Issue by nblackburn** [Astray-git/vue-dragula#12]
When creating multiple directives with the same bag, the following error is thrown...

``` html
<div class="wrapper">
  <div class="container" v-dragula="colOne" bag="first-bag">
    <div v-for="text in colOne" @click="onClick">{{text}} [click me]</div>
  </div>
  <div class="container" v-dragula="colTwo" bag="first-bag">
    <div v-for="text in colTwo">{{text}}</div>
  </div>
</div>
```

> Uncaught Error: Bag named: "test" already exists.

**Comment by Astray-git** [comment]
Can you show me your `option` code?  (In which life cycle hook)

**Comment by nblackburn** [comment]
Sure, the code lives on the `ready` event and looks like this...

``` html
<div v-el:filters v-dragula="filters" bag="filters"></div>
<div v-el:attached v-dragula="attached" bag="filters"></div>
```

``` javascript
ready() {

    let filterContainer = this.$els.filters;
    let activeContainer = this.$els.attached;

    Vue.vueDragula.options('filters', {

        removeOnSpill: true,
        revertOnSpill: true,
        direction: 'vertical',

        copy: (element, origin) => {
            return origin === filterContainer;
        },

        accepts: (element, target) => {
            return target === activeContainer;
        }
    });
}
```

**Comment by Astray-git** [comment]
Try move them to `created` hook, so bags set up before directive `bind`.

**Comment by nblackburn** [comment]
That has got rid of the error but because the DOM isn't ready in this event, can't access the element bindings used in the `copy` and `accepts` properties.

**Comment by Astray-git** [comment]
You can put `filterContainer` as a component global variable, update it on `ready`, or use id to match containers.

**Comment by kristianmandrup** [comment]
I had the same issue. I don't understand the concept of bags from the docs. Also, how can we NOT get the error if there is only one global service and reuse of the same bag name on multiple directives instances!?

**Comment by kristianmandrup** [comment]
@nblackburn What is your status? Let's work together on a better solution and good patterns for Vue 2.x. This is a jungle!!!

**Comment by kristianmandrup** [comment]
I had the same issue. I moved it to `mounted` but not sure where in the lifecycle is correct.
I added a way to have multiple services, but then...!?!? I'm lost! 
I'm newbie on Vue still :( Why must web dev still be so hard in 2016...

**Comment by kristianmandrup** [comment]
"You can put filterContainer as a component global variable, update it on ready, or use id to match containers." WTF!? Don't understand this gibbidy-gack! 
@Astray-git Please show an example for Vue 2 :)

**Comment by Astray-git** [comment]
Call `option` in a component will add a new bag. In directive bind hook It try to add a new bag if there's no existing one.
So the correct order is config dragula instance then initalize the directive (set options before directive 'bind'), you can try the `created` hook. (which is almost a same hook in both Vue 1.x and 2)

**Comment by kristianmandrup** [comment]
"Call option in a component will add a new bag" ??? what `option`?

Ah, you mean `Vue.vueDragula.options` which calls `setOptions` on `DraculaService` instance, which calls `add(bagName, drake)` which adds a named bag with a drake (dragula) instance

" then initalize the directive" - How can I control when directive is initialised? I thought that was automatic!!

Thanks anyways... Still not helping much. Wish you could make a working Vue2 example/demo. I think a lot of us are sad/frustrated having to 2nd guess how to make it work in Vue2.
I've been working all day (and last evening) trying to make sense of it...

**Comment by kristianmandrup** [comment]
OK, we need to clearly understand the [life cycle](https://apimirror.com/vue~2/api/index#created) but still I don't understand what must go where in these hooks.

**Comment by Astray-git** [comment]
In vue instance's `created` hook directive are not initialized yet. we call `Vue.vueDragula.options` here, a bag is ready.
Later in directive's `bind` hook, we add models etc into the dragula instance stored in bag.

**Comment by kristianmandrup** [comment]
So far, I could never get your example app to work in Vue 2, exactly because you call `options` with `copy: true` for `third-bag` and it throws exception that `third-bag` is already taken.

[third-bag options](https://github.com/Astray-git/vue-dragula/blob/master/example/example.js#L35) and it is in `created` hook:

``` js
  created: function () {
    Vue.vueDragula.options('third-bag', {
      copy: true
    })
  },
```

I've been fighting this exact problems for the past many hours...

**Comment by kristianmandrup** [comment]
When you bind each directive, you end up calling `service.add(name, drake)` where name is the name of the bag, ie.

``` html
<div class="wrapper">
  <div class="container" v-dragula="colOne" bag="first-bag">
    <!-- with click -->
    <div v-for="text in colOne" @click="onClick">{{text}} [click me]</div>
  </div>
  <div class="container" v-dragula="colTwo" bag="first-bag">
    <div v-for="text in colTwo">{{text}}</div>
  </div>
</div>
```

But any call to add will cause a check on existing bag name, and so on the second call (ie second binding with `bag="same-name"` it will always cause an exception as far as I can tell!? No matter where in the life cycle?

``` js
    let bag = this.find(name)
    if (bag) {
      throw new Error('Bag named: "' + name + '" already exists for this service')
    }
```

**Comment by kristianmandrup** [comment]
And this is explained in the docs/Readme!? I guess I missed that...

**Comment by Astray-git** [comment]
https://github.com/Astray-git/vue-dragula/blob/master/src/vue-dragula.js#L33-L38
in `bind` if there's already a bag with the same name, `add` will not be called.

**Comment by kristianmandrup** [comment]
I finally get it now :) Thanks! Was not clear you HAD to use options in `created` hook in order for directives to work correctly. Has taken me all this effort to understand this key insight. 
A bit too brittle for my taste. Thanks anyways ;)

**Comment by kristianmandrup** [comment]
So I assume the problem is that `bind` is asynchronous!? and so there can be race condition!?
So the directive both end up calling `service.add` without knowing or noticing that a bag is already there... this is the conflict!?

**Comment by kristianmandrup** [comment]
I think the bags concept is very confusing. What you are trying to achieve is a map of `name: dragula instance (drake)` so just use an Object, also much faster lookup that what you are using for `find` iterating through the list manually looking for object with matching name.

**Comment by kristianmandrup** [comment]
I finally "GOT IT" and now my extension of vue-dragula now has a beautiful, intuitive API to manage all this. Time to play and have fun :)

**Comment by Astray-git** [comment]
This should be solved in version 2.0.0
